superset/utils/date_parser.py

```diff
@@ -263,6 +264,47 @@ def get_since_until(  # pylint: disable=too-many-arguments,too-many-locals,too-m
         _since = _since if _since is None else (_since - time_delta)
         _until = _until if _until is None else (_until - time_delta)
 
+    if instant_time_comparison_range:
+        # This is only set using the new time comparison controls
+        # that is made available in some plugins behind the experimental
+        # feature flag.
+        # pylint: disable=import-outside-toplevel
+        from superset import feature_flag_manager
+
+        if feature_flag_manager.is_feature_enabled("CHART_PLUGINS_EXPERIMENTAL"):
+            time_unit = ""
+            delta_in_days = None
+            if instant_time_comparison_range == InstantTimeComparison.YEAR:
+                time_unit = "YEAR"
+            elif instant_time_comparison_range == InstantTimeComparison.MONTH:
+                time_unit = "MONTH"
+            elif instant_time_comparison_range == InstantTimeComparison.WEEK:
+                time_unit = "WEEK"
+            elif instant_time_comparison_range == InstantTimeComparison.INHERITED:
+                delta_in_days = (_until - _since).days if _since and _until else None
```

Comment:
I just understood what is "inherited" meaning, which is date time difference between since and until. So, for more general use we might create a new function in our custom date time expression e.g. `DATEDIFF([since], [until], [unit])`. 

Then we could generate a expression like below:

`DATEADD(DATETIME(...), -DATEDIFF(..., ..., DAY), DAY) : DATEADD(DATETIME(...), -DATEDIFF(..., ..., DAY), DAY)`

After that, we can move all of the backend hard-code into frontend.

Reply:
BTW, the PR is totally work for me, we could improve in the follow-up PR>

Reply:
About doing the expression in the frontend, I still consider it's best to do it in the backend because the backend processes the right `datetime` no matter what value the filter has.

If the frontend wants to calculate the right `datetime` it will fail because as stated before, an expression that for example has the filter with value: `last year` will get an invalid `datetime` if you wrap it with `DATETIME` and making the frontend calculate the `"right" datetime` would lead to duplicating logic that already works well in the backend.

You could for example do a quick test of setting a filter's value to: `last year` vs `DATETIME('last year')` and see what I'm saying about the difference in both ways.

TLDR; Yes, we could introduce a `DATEDIFF` but IMO we should keep using the backend for building these ranges.

Reply:
There are some legacy logics and concepts from the initial Superset; let me explain them before, and then you might get more context of the time picker and the original design of the time_range.

First of all, the time_range has two forms:

1. plain text, e.g. `time_range = "last year"`
2. plain text with colon and spaces. e.g. `time_range = " 2021-03-01 : dateadd(datetime("tomorrow"), 1, day)"`

The plain text form is applied in "Last" and "Previous" Frame in the Time Picker, these frame will generate phrases like 'last year', 'last month', 'previous calendar month', using natural English languages to define a time period; The plain text with colon and spaces is more straightforward, it directly define a `since : until` time period.

Notice that these two forms of time_range will convert to the uniform "since : until" in the backend,
- if `Last / Next` in the beginning of the [time_range](https://github.com/apache/superset/blob/54037972f2f56ec86a613e6684e8bbe20cafcf50/superset/utils/date_parser.py#L185-L189), the compiled the expression will be [here](https://github.com/apache/superset/blob/54037972f2f56ec86a613e6684e8bbe20cafcf50/superset/utils/date_parser.py#L210-L219)
- if `Previous` in the beginning of the time_range, the compiled the expression will be [here](https://github.com/apache/superset/blob/54037972f2f56ec86a613e6684e8bbe20cafcf50/superset/utils/date_parser.py#L191-L208).

From the [codes](https://github.com/apache/superset/blob/54037972f2f56ec86a613e6684e8bbe20cafcf50/superset/utils/date_parser.py#L214), 
- If you choose`Last year` in the Last frame in Time Picker, the compiled expression should be: `DATEADD(DATETIME('today'), -1, year) : `, means that relative today in the last year.
- If you input `datetime('last year')` in the custom Frame, means that the first day of the last year, it's not equal to `Last year` choose in the Last frame.

So if you want to generate a time shift in the frontend, you need to 
1. implement DATEDIFF function in the [datetime_parser](https://github.com/apache/superset/blob/54037972f2f56ec86a613e6684e8bbe20cafcf50/superset/utils/date_parser.py#L440), in order to calculate time delta.
2. use `DATEADD(DATEADD(DATETIME('today'), -1, year), -1, year) : ` if you want to generate a relative date of `Last year` time_range.

In summary, we don't need to hard code in get_since_until function, we just need to find out what tools is existing.

Why I said so much in this PR since I found this project we've introduced lots of inconsistent design and hard code anywhere for some temporally request.